#333: change color scale and other graphical improvements

ppebay opened the pull request:
Resolves #333

ppebay commented:
@lifflander this is still in draft mode but a first review would be great anyway. Thanks

ppebay commented:
Categorical data detection and rendering

<img width="750" alt="Screen Shot 2022-12-24 at 12 10 53 AM" src="https://user-images.githubusercontent.com/17439260/209412516-e3b44030-51bd-46a1-81d6-14b5fb74b6c8.png">
<img width="739" alt="Screen Shot 2022-12-24 at 12 11 06 AM" src="https://user-images.githubusercontent.com/17439260/209412522-58a322b3-0d7a-4ee9-a490-0fa7610481d6.png">

ppebay commented:
Arbitrary object QOI (but scaling remains based on object load irrespective of chosen object QOI)


<img width="742" alt="Screen Shot 2022-12-24 at 12 16 31 AM" src="https://user-images.githubusercontent.com/17439260/209412758-99423200-2fc9-4f41-b5e7-a41e3d6e6702.png">

<img width="739" alt="Screen Shot 2022-12-24 at 12 11 06 AM" src="https://user-images.githubusercontent.com/17439260/209412786-2a465d13-7a90-4b30-a775-e29ddf247139.png">

ppebay commented:
Categorical labeling with annotations:
<img width="742" alt="Screen Shot 2022-12-24 at 12 16 31 AM" src="https://user-images.githubusercontent.com/17439260/209416537-a458dae3-6f7c-4b5e-a5e7-d89165079786.png">

ppebay commented:
Better categorical annotations:

<img width="742" alt="Screen Shot 2022-12-24 at 12 16 31 AM" src="https://user-images.githubusercontent.com/17439260/209416565-2ec2c57a-bd2d-4251-8801-0c8ba7b19069.png">

cz4rs commented on src/lbaf/Model/lbsPhase.py:
```diff
@@ -79,9 +79,19 @@ def get_number_of_objects(self):
+
+        """ Return all objects belonging to phase."""
+
+        objects = []
+        for p in self.__ranks:
+            objects += p.get_objects()
```
```suggestion
        objects = [p.get_objects() for p in self.__ranks]
```

ppebay replied:
Actually no, because here `+` is used to perform a **list concatenation**; the suggested change results in creating a list of lists (which is not what we want).

cz4rs commented on src/lbaf/IO/lbsMeshBasedVisualizer.py:
```diff
@@ -36,28 +37,27 @@ def __init__(
-        if not isinstance(qoi, list) or not len(qoi) or not (
-            qoi_name := qoi[0]) or not isinstance(qoi_name, str):
+        # Make sure that quantity of interest names were passed
+        if not isinstance(qoi_request, list) or not len(qoi_request) > 2 or not (
+            rank_qoi := qoi_request[0]) or not isinstance(
+            rank_qoi, str) or not (object_qoi := qoi_request[2]) or not isinstance(
+            object_qoi, str):
```
This is quite hard to read. Can you extract it into a separate helper (returning `bool`) or split it up into named boolean variables?

cz4rs commented on src/lbaf/IO/lbsMeshBasedVisualizer.py:
```diff
@@ -512,43 +568,43 @@ def create_rendering_pipeline(
+        sqrtL_out.GetPointData().SetActiveScalars("Migratable")
 
         # Glyph sentinel and migratable objects separately
         glyph_actors, glyph_mapper = [], None
```
```local variable 'glyph_actors' is assigned to but never used```

cz4rs commented on src/lbaf/IO/lbsMeshBasedVisualizer.py:
```diff
@@ -92,20 +92,42 @@ def __init__(
-                    # Update load range when necessary
-                    load = o.get_load()
-                    if load > self.__load_range[1]:
-                        self.__load_range[1] = load
-                    if load < self.__load_range[0]:
-                        self.__load_range[0] = load
+            for o in p.get_objects():
+                # Update maximum object load as needed
+                if (ol := o.get_load()) > self.__object_load_max:
+                    self.__object_load_max = ol
+
+                # Retain all QOI values while support remains small
+                oq = getattr(o, f"get_{object_qoi}")()
+                if store_all:
+                    oq_all.add(oq)
+                    if len(oq_all) > 20:
+                        # Do not store QOI values if support is too large
+                        oq_all = None
+                        store_all = False
+
+                # Update extrema
+                if oq < oq_min:
+                    oq_min = oq
+                if oq > oq_max:
+                    oq_max = oq
+
+        # Store either range or support
+        if store_all:
+            self.__object_qoi_range = oq_all
+            self.__logger.info(
+                f"\t{self.__object_qoi} has {len(self.__object_qoi_range)} distinct values")
+        else:
+            self.__object_qoi_range = (oq_min, oq_max)
+            self.__logger.info(
+                f"\t{self.__object_qoi} range: [{self.__object_qoi_range[0]:.4g}; {self.__object_qoi_range[1]:.4g}]")
```
`__init__` is now over 200 lines long. Can you extract this into a separate method?

ppebay commented on src/lbaf/Applications/LBAF_app.py:
```diff
@@ -162,19 +162,30 @@ def parse_conf_file(self):
+                self.object_jitter = viz["object_jitter"]
+                self.rank_qoi = viz["rank_qoi"]
+                self.object_qoi = viz["object_qoi"]
+            except:
```
Yes but at this point we just want to error out brutally, in case **any** exception is encountered. Not trying to be smart about the exception type.

ppebay commented:
All changes requested by @cz4rs either done or justified when not done. However GH still indicates there are unaddressed comments -- which I can't see. Re-requesting review

ppebay reviewed (commented):
All changes requested by @cz4rs either made, or justification added when not made.
GH still blocking merge nonetheless, looks like a CI bug.

ppebay commented:
@cz4rs all comments addressed --- please check in your review what's still blocking the merge. Thanks!

cz4rs reviewed (approved):
Looks good to me!

cz4rs commented:
Note: the code looks good and the visualization looks fine to me, but you might still want to double check the vis~ with others.

ppebay commented:
@lifflander @nlslatt in case you have time to do a post-merge review that would be great, thanks
